File: action-scheduler-stuck-in-progress/node/audit-stuck-actions.js
/**
 * Find Action Scheduler actions stuck on in-progress and decide how to clear them.
 *
 * An action normally moves from pending, to in-progress, to complete within seconds.
 * When the PHP worker that claimed an action dies mid run (a timeout, an out of
 * memory kill, a fatal error), the action is left on in-progress forever. Action
 * Scheduler's own claim lock then treats that slot as busy, so the next run of that
 * group or hook can stall behind it, and the queue backs up.
 *
 * This script does not touch wp_actionscheduler_actions directly. It uses the
 * WooCommerce REST API to read the order that a stuck subscription renewal or
 * payment action points to (order id comes from a small JSON export of stuck
 * actions), asks Stripe for the truth about the PaymentIntent on that order, and
 * decides one of four outcomes: complete_order, reset_action, wait, or investigate.
 *
 * Read only by default (DRY_RUN=true). Run on a schedule.
 *
 * Guide: https://www.allanninal.dev/woocommerce/action-scheduler-stuck-in-progress/
 */
import Stripe from "stripe";
import { readFile } from "node:fs/promises";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const STUCK_AFTER_MINUTES = Number(process.env.STUCK_AFTER_MINUTES || 30);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const PAID_STATUSES = new Set(["processing", "completed"]);
const IN_FLIGHT_INTENT_STATUSES = new Set(["requires_action", "requires_confirmation", "processing"]);
const FAILED_INTENT_STATUSES = new Set(["requires_payment_method", "canceled"]);

export function intentIdOf(order) {
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  const tid = order.transaction_id;
  return tid && tid.startsWith("pi_") ? tid : null;
}

/**
 * Pure decision function. No I/O. action is an object with at least status and
 * ageMinutes. order and intent may be null. Returns [verdict, reason].
 */
export function decide(action, order, intent) {
  if (action.status !== "in-progress") return ["skip", "action is not in-progress"];
  if ((action.ageMinutes || 0) < STUCK_AFTER_MINUTES) return ["wait", "action has not been stuck long enough yet"];
  if (!order) return ["investigate", "action points to an order that cannot be found"];
  if (intent && IN_FLIGHT_INTENT_STATUSES.has(intent.status)) {
    return ["investigate", "Stripe shows the payment still in flight"];
  }
  if (PAID_STATUSES.has(order.status)) return ["reset_action", "order is already paid, the action is just stale"];
  if (intent && intent.status === "succeeded") return ["complete_order", "Stripe succeeded but the order was never updated"];
  if (!intent || FAILED_INTENT_STATUSES.has(intent.status)) {
    return ["reset_action", "no successful charge behind this attempt, safe to retry"];
  }
  return ["investigate", "unclear Stripe state, needs a human look"];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function getIntent(intentId) {
  if (!intentId) return null;
  try {
    return await stripe.paymentIntents.retrieve(intentId);
  } catch {
    return null;
  }
}

async function completeOrder(orderId, intent) {
  const chargeId = intent.latest_charge || intent.id;
  await woo(`/orders/${orderId}`, {
    method: "PUT",
    body: JSON.stringify({ status: "processing", transaction_id: chargeId }),
  });
  await woo(`/orders/${orderId}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Recovered from a stuck Action Scheduler action. Stripe PaymentIntent ` +
            `${intent.id} had already succeeded. Marked processing by the auditor.`,
    }),
  });
}

async function noteStuckAction(orderId, reason) {
  await woo(`/orders/${orderId}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Action Scheduler action for this order was stuck on in-progress: ` +
            `${reason}. Flagged for a reset by the auditor.`,
    }),
  });
}

/**
 * Read a small JSON export of stuck actions. Each row looks like:
 * { "actionId": 4821, "status": "in-progress", "ageMinutes": 55, "orderId": 9321 }
 * Produce this with:
 *   wp action-scheduler action list --status=in-progress --format=json
 * then add ageMinutes and orderId per hook args, or adapt this loader to your
 * own store's export shape.
 */
async function stuckActionsFromExport(path) {
  const raw = await readFile(path, "utf8");
  return JSON.parse(raw);
}

export async function run(exportPath = "stuck.json") {
  let handled = 0;
  for (const action of await stuckActionsFromExport(exportPath)) {
    const orderId = action.orderId;
    const order = orderId ? await woo(`/orders/${orderId}`) : null;
    const intent = order ? await getIntent(intentIdOf(order)) : null;
    const [verdict, reason] = decide(action, order, intent);
    if (verdict === "skip" || verdict === "wait") continue;
    console.log(
      `Action ${action.actionId} (order ${orderId}): ${reason} -> ${DRY_RUN ? "would act" : "acting"}`
    );
    if (!DRY_RUN) {
      if (verdict === "complete_order") await completeOrder(orderId, intent);
      else if (verdict === "reset_action" || verdict === "investigate") await noteStuckAction(orderId, reason);
    }
    handled++;
  }
  console.log(`Done. ${handled} stuck action(s) ${DRY_RUN ? "to handle" : "handled"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run(process.argv[2]).catch((e) => { console.error(e); process.exit(1); });
}

File: action-scheduler-stuck-in-progress/python/audit_stuck_actions.py
"""Find Action Scheduler actions stuck on in-progress and decide how to clear them.

An action normally moves from pending, to in-progress, to complete within seconds.
When the PHP worker that claimed an action dies mid run (a timeout, an out of memory
kill, a fatal error), the action is left on in-progress forever. Action Scheduler's
own claim lock then treats that slot as busy, so the next run of that group or hook
can stall behind it, and the queue backs up.

This script does not touch wp_actionscheduler_actions directly. It uses the
WooCommerce REST API to read the order that a stuck subscription renewal or payment
action points to (order id is taken from the action's hook args, passed in on the
command line or from a small JSON export), asks Stripe for the truth about the
PaymentIntent on that order, and decides one of four outcomes:

  - "complete_order": Stripe says the payment succeeded. Mark the order processing
    and add a note. The stuck action can be marked complete in wp-admin or with
    `wp action-scheduler action update --id=<id> --status=complete`.
  - "reset_action": Stripe never took a real payment for this attempt (no intent,
    or the intent failed or is still requiring action). Safe to reset the action
    back to pending so it can be retried, since nothing was charged.
  - "wait": the action has not been stuck long enough yet to act on. Actions can
    briefly sit on in-progress during a normal, slow run.
  - "investigate": Stripe shows a PaymentIntent that requires_action or is
    processing. Money is in flight. Do not touch the order or the action yet.

Read only by default (DRY_RUN=true). Run on a schedule, for example every 15
minutes, well past STUCK_AFTER_MINUTES.
"""
import os
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("audit_stuck_actions")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
STUCK_AFTER_MINUTES = int(os.environ.get("STUCK_AFTER_MINUTES", "30"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

PAID_STATUSES = {"processing", "completed"}
IN_FLIGHT_INTENT_STATUSES = {"requires_action", "requires_confirmation", "processing"}
FAILED_INTENT_STATUSES = {"requires_payment_method", "canceled"}


def intent_id_of(order):
    """The saved Stripe PaymentIntent id, from meta _stripe_intent_id or transaction_id."""
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    tid = order.get("transaction_id")
    return tid if tid and tid.startswith("pi_") else None


def decide(action, order, intent):
    """Pure decision function. No I/O. action is a dict with at least
    status and age_minutes. order and intent may be None.
    Returns (verdict, reason).
    """
    if action.get("status") != "in-progress":
        return ("skip", "action is not in-progress")
    if action.get("age_minutes", 0) < STUCK_AFTER_MINUTES:
        return ("wait", "action has not been stuck long enough yet")
    if order is None:
        return ("investigate", "action points to an order that cannot be found")
    if intent is not None and intent.get("status") in IN_FLIGHT_INTENT_STATUSES:
        return ("investigate", "Stripe shows the payment still in flight")
    if order.get("status") in PAID_STATUSES:
        return ("reset_action", "order is already paid, the action is just stale")
    if intent is not None and intent.get("status") == "succeeded":
        return ("complete_order", "Stripe succeeded but the order was never updated")
    if intent is None or intent.get("status") in FAILED_INTENT_STATUSES:
        return ("reset_action", "no successful charge behind this attempt, safe to retry")
    return ("investigate", "unclear Stripe state, needs a human look")


def get_order(order_id):
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def get_intent(intent_id):
    if not intent_id:
        return None
    try:
        return stripe.PaymentIntent.retrieve(intent_id)
    except stripe.error.InvalidRequestError:
        return None


def complete_order(order_id, intent):
    charge_id = intent.get("latest_charge") or intent["id"]
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}",
        json={"status": "processing", "transaction_id": charge_id},
        auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/notes",
        json={"note": f"Recovered from a stuck Action Scheduler action. Stripe PaymentIntent "
                      f"{intent['id']} had already succeeded. Marked processing by the auditor."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def note_stuck_action(order_id, reason):
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/notes",
        json={"note": f"Action Scheduler action for this order was stuck on in-progress: "
                      f"{reason}. Flagged for a reset by the auditor."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def stuck_actions_from_export(path):
    """Read a small JSON export of stuck actions. Each row looks like:
    {"action_id": 4821, "status": "in-progress", "age_minutes": 55, "order_id": 9321}
    Produce this with:
    wp action-scheduler action list --status=in-progress --format=json > stuck.json
    then add age_minutes and order_id per hook args before feeding it in, or adapt
    this loader to your own store's export shape.
    """
    import json
    with open(path) as f:
        return json.load(f)


def run(export_path):
    handled = 0
    for action in stuck_actions_from_export(export_path):
        order_id = action.get("order_id")
        order = get_order(order_id) if order_id else None
        intent = get_intent(intent_id_of(order)) if order else None
        verdict, reason = decide(action, order, intent)
        if verdict in ("skip", "wait"):
            continue
        log.info(
            "Action %s (order %s): %s -> %s",
            action.get("action_id"), order_id, reason,
            "would act" if DRY_RUN else "acting",
        )
        if not DRY_RUN:
            if verdict == "complete_order":
                complete_order(order_id, intent)
            elif verdict in ("reset_action", "investigate"):
                note_stuck_action(order_id, reason)
        handled += 1
    log.info("Done. %d stuck action(s) %s.", handled, "to handle" if DRY_RUN else "handled")


if __name__ == "__main__":
    import sys
    run(sys.argv[1] if len(sys.argv) > 1 else "stuck.json")
